src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
+				if (isCumulative) {
+					meterReading1 = rows[index - 2 * readingRepetition][0]; // use this if regular
+					meterReading2 = rows[index - readingRepetition][0];
+					meterReading = meterReading2 - meterReading1; // use this ifCumulative
+				}
+				else {
+					meterReading = rows[index - readingRepetition][0];
+				}
+				// Reject negative readings
+				if (isCumulative && meterReading1 < 0) {
+					// If the value isCumulative and the first reading is negative reject unless cumulativeReset is true
+					log.error(`DETECTED A NEGATIVE VALUE WHILE HANDLING CUMULATIVE READINGS FROM METER ${meterID}, ` +
+						`ROW ${index - readingRepetition}. REJECTED ALL READINGS`);
+					return [];
+				}
+				// To handle cumulative readings that resets at midnight
+				if (meterReading < 0) {
+					// if meterReading is negative and cumulative check that the times fall within an acceptable reset range
+					if (handleCumulativeReset(cumulativeReset, resetStart, resetEnd, startTimestamp)) {
+						meterReading = meterReading2;
+					}
+					else {
+						//cumulativeReset is not expected so throw an error
+						errMsg = ('A negative meterReading has been detected but either cumulativeReset is not enabled, or the start time and end time of this reading is out of the reset range.')
+						readingOK = false;
+					}
+				}
+				if (readingOK) {
+					if (Math.abs(prevReading.endTimestamp.diff(prevReading.startTimestamp) - endTimestamp.diff(startTimestamp)) 
+						> Tlen && !isFirst(prevReading.endTimestamp)) {
+						errMsg = 'The previous reading has a different time length than the current reading.';
+					}
+					currentReading = new Reading(meterID, meterReading, startTimestamp, endTimestamp);
+					result.push(currentReading);
+				}
+			}
+			if (!readingOK) {
+				// An error occurred, for now log it, let the user know and continue
+				log.error(errMsg);
+				readingOK = true;
+				if (isCumulative && !onlyEndTime || onlyEndTime && !isCumulative || isCumulative && onlyEndTime) {
```

Comment:
I've added this comment to make it clear why this logic exists.

/*
If the data is cumulative then regardless of if it comes with end timestamps only or both end timestamps and start timestamps the first reading ever should become the previous reading. This is necessary because there are no previous readings in the db yet so we must drop the first point ever and use this first point as the first previous reading in order to begin calculating net readings since the data is cumulative.

If the data is not cumulative but there are only end timestamps then we still must drop the first reading ever in order to use that reading as the start timestamp for all following readings. 
*/

I think now though after writing this comment out, it makes more sense to instead use isFirst() to check the first reading since this is the only case this logic is necessary. 

What do you think?